lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,171 @@
+      discover(context: context, logger: logger) unless @resources
+
+      type = definition["kind"]
+      group, _, version = definition['apiVersion'].rpartition('/')
+
+      resource_class = @resources.dig(group, type, version) if @resources
+      return super unless resource_class
+
+      opts = { namespace: namespace, context: context, definition: definition, logger: logger }
+      resource_class.new(**opts)
+    end
+
+    def self.discover_tpr(client)
+      return unless client.respond_to? :get_third_party_resources
+      resources = client.get_third_party_resources
+      resources.each do |res|
+        type, _, group = res.metadata.name.partition('.')
+        # TPR API supports multiple versions for a single resource :(
+        res.versions.each do |version|
+          discovered(group: group,
+                     type: type,
+                     version: version.name,
+                     annotations: res.metadata.annotations)
+        end
+      end
+    end
+
+    def self.discover_crd(client)
+      return unless client.respond_to? :get_custom_resource_definitions
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        discovered(group: res.spec.group,
+                   type: res.spec.names.kind,
+                   version: res.spec.version,
+                   annotations: res.metadata.annotations)
+      end
+    end
+
+    def self.discovered(group:, type:, version:, annotations:)
+      resource_class = Class.new(self) do
+        type = type.capitalize # TPRs are inconsistent about capitalization
+        @group = group
+        @type = type
+        @version = version
+        @prunable = DiscoverableResource.parse_bool(annotations[PRUNABLE_ANNOTATION])
+        @predeploy = DiscoverableResource.parse_bool(annotations[PREDEPLOY_ANNOTATION])
+        @timeout = DiscoverableResource.parse_timeout(type, annotations[TIMEOUT_ANNOTATION])
+
+        status_field = annotations[STATUS_FIELD_ANNOTATION]
+        success_status = annotations[STATUS_SUCCESS_ANNOTATION]
+
+        define_method 'deploy_succeeded?' do
+          getter = "get_#{type.downcase}"
+          @client ||= DiscoverableResource.kubeclient(context: @context, resource_class: self.class)
+          raw_json = @client.send(getter, @name, @namespace, as: :raw)
+          query_path = JsonPath.new(status_field)
+          current_status = query_path.first(raw_json)
+          current_status == success_status
+        end if status_field && success_status
+      end
+
+      add_resource(resource_class)
+    end
+
+    def self.add_resource(resource_class)
+      group = resource_class.group
+      type = resource_class.type
+      version = resource_class.version
+      @resources ||= {}
+      @resources[group] ||= {}
+      @resources[group][type] ||= {}
+      @resources[group][type][version] = resource_class
+    end
+
+    def self.parse_bool(value)
```

Comment:
Kubernetes also uses `yes` for the value of a boolean annotation. These annotations are added to persistent volume claims once they are bound.

```
pv.kubernetes.io/bind-completed=yes
pv.kubernetes.io/bound-by-controller=yes
```

This should also look for `yes`.

Reply:
lets do ON, yes, true and 1 wdyt?

Reply:
The value needs to be a string so "1" might get confusing.

Reply:
FWIW this is what Rails (well, ActiveModel) does: https://github.com/rails/rails/blob/47eadb68bfcae1641b019e07e051aa39420685fb/activemodel/lib/active_model/type/boolean.rb#L17

Reply:
Switched to ActiveModel, thanks!